### ISSUE TITLE
Windows build?

### ISSUE THREAD

**Issue by BatmanAoD** [oilshell/oil#122]
Does Oil build on Windows? If so, does it require MinGW with standard NIX-y tools (`make`, `cc`, etc), or is there a way to build it with Visual Studio components?

(I tried looking for "Windows" and "MinGW" in the issues list, but didn't see anything.)

**Comment by andychu** [comment]
I haven't tried it myself, but someone told me it works fine on Windows subsystem for LInux:

https://github.com/oilshell/oil/wiki/Oil-Deployments

(Windows 10 I believe).

I guess that doesn't use MinGW, but it's like an Ubuntu distro on top of a Windows kernel.

I think you need the kernel components too?  It has to be something like Cygwin I think, not MinGW because that is just tools and no kernel emulation?

So basically pre-Windows 10 I would expect you need Cygwin, but I haven't tried it.

**Comment by BatmanAoD** [comment]
Yeah, getting stuff to work on the Subsystem for Linux *should* be pretty trivial! But that doesn't actually mean that the resulting binary will actually be usable outside the subsystem itself (i.e. you can't just invoke it from the Windows command shell).

I more or less despise Cygwin, so I probably won't look into building it that way, but I may at some point look into whether there's some way to compile with native Windows tools. From a brief glance, it looks like the main thing that's needed is `libc`. MinGW might provide the necessary translations to the corresponding MSVCRT methods, or maybe we need some platform-specific `#ifdef`s. Either way, I don't think this should be insurmountable!

**Comment by andychu** [comment]
I'm not sure how you would do fork() and exec() that way.  MSVCRT doesn't have those as far as I know.  They're not really part of ANSI C but POSIX as far as I know.

I think you would end up just writing a tiny Cygwin...

Anyway since this isn't that specific a task, let's discuss any further on #oil-dev on Zulip:

https://oilshell.zulipchat.com/#narrow/stream/121539-oil-dev

(you can log in with Github)

Others might chime in.  Happy to discuss more.

Although I suppose there is a native build of bash for git on Windows?  I haven't ever used that but that's what I heard.  I'm not sure how it works.

**Comment by 89z** [comment]
I am looking for a PowerShell replacement, so I support this.

**Comment by BatmanAoD** [comment]
@89z If you like PowerShell's somewhat "object-oriented" pipe rules, may I suggest taking a look at [Nushell](https://github.com/nushell/nushell/)?

**Comment by 89z** [comment]
Thanks for the suggestion. NuShell looks like a good option, but they have some broken stuff too, so I am waiting for the next release.
